File: libs/power-log-parser/src/lib/parsers/battlegrounds-hero-selection-parser.ts
import { CardType, GameTag, Mulligan, Zone } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider } from '../game-event';
import { Node, NodeType, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class BattlegroundsHeroSelectionParser implements ActionParser {
	readonly ParserName = 'BattlegroundsHeroSelectionParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, helper: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = helper;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === NodeType.TagChange &&
			(node.Object as TagChange).Name === (GameTag.MULLIGAN_STATE as number) &&
			(node.Object as TagChange).Value === (Mulligan.INPUT as number)
		);
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tagChange = node.Object as TagChange;
		const playerId = this.StateFacade.LocalPlayer!.PlayerId;
		const fullEntities = [...this.GameState.CurrentEntities.values()]
			.filter((data) => data.GetEffectiveController() === playerId)
			.filter((data) => data.GetTag(GameTag.CARDTYPE) === (CardType.HERO as number))
			.filter((data) => data.GetTag(GameTag.ZONE) === (Zone.HAND as number))
			.filter(
				(data) =>
					data.GetTag(GameTag.BACON_HERO_CAN_BE_DRAFTED) === 1 ||
					data.GetTag(GameTag.BACON_SKIN) === 1,
			);
		fullEntities.sort((a, b) => a.GetTag(GameTag.ZONE_POSITION) - b.GetTag(GameTag.ZONE_POSITION));
		if (fullEntities.length > 0) {
			return [
				GameEventProvider.Create(
					tagChange.TimeStamp,
					'BATTLEGROUNDS_HERO_SELECTION',
					() => {
						if (!this.StateFacade.IsBattlegrounds()) {
							return null;
						}
						return {
							Type: 'BATTLEGROUNDS_HERO_SELECTION',
							Value: {
								Options: fullEntities.map((entity) => ({
									CardId: entity.CardId,
									EntityId: entity.Id,
								})),
							},
						};
					},
					true,
					node,
				),
			];
		}
		return null;
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/BattlegroundsHeroSelectionParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System.Linq;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using static HearthstoneReplays.Events.CardIds;

namespace HearthstoneReplays.Events.Parsers
{
    public class BattlegroundsHeroSelectionParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public BattlegroundsHeroSelectionParser(ParserState ParserState, StateFacade helper)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = helper;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(TagChange)
                && (node.Object as TagChange).Name == (int)GameTag.MULLIGAN_STATE
                && (node.Object as TagChange).Value == (int)Mulligan.INPUT;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            var playerId = StateFacade.LocalPlayer.PlayerId;
            var fullEntities = GameState.CurrentEntities.Values
                .Where(data => data.GetEffectiveController() == playerId)
                .Where(data => data.GetTag(GameTag.CARDTYPE) == (int)CardType.HERO)
                .Where(data => data.GetTag(GameTag.ZONE) == (int)Zone.HAND)
                .Where(data => data.GetTag(GameTag.BACON_HERO_CAN_BE_DRAFTED) == 1 || data.GetTag(GameTag.BACON_SKIN) == 1)
                .ToList();
            fullEntities.Sort((a, b) => a.GetTag(GameTag.ZONE_POSITION) - b.GetTag(GameTag.ZONE_POSITION));
            Logger.Log("will consider hero selection event", "" + fullEntities.Count);
            if (fullEntities.Count > 0)
            {
                Logger.Log("will emit hero selection event", "");
                return new List<GameEventProvider> { GameEventProvider.Create(
                   tagChange.TimeStamp,
                   "BATTLEGROUNDS_HERO_SELECTION",
                   () => {
                       if (!StateFacade.IsBattlegrounds())
                       {
                           return null;
                       }
                       return new GameEvent
                       {
                           Type = "BATTLEGROUNDS_HERO_SELECTION",
                           Value = new
                           {
                               Options = fullEntities.Select(entity => new {
                                   CardId = entity.CardId,
                                   EntityId = entity.Id,
                               }).ToList(),
                           }
                       };
                   },
                   true,
                   node
                )};
            }
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
